File: sonar-server/src/main/webapp/WEB-INF/app/models/filter_column.rb
#
# Sonar, entreprise quality control tool.
# Copyright (C) 2008-2012 SonarSource
# mailto:contact AT sonarsource DOT com
#
# Sonar is free software; you can redistribute it and/or
# modify it under the terms of the GNU Lesser General Public
# License as published by the Free Software Foundation; either
# version 3 of the License, or (at your option) any later version.
#
# Sonar is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# Lesser General Public License for more details.
#
# You should have received a copy of the GNU Lesser General Public
# License along with Sonar; if not, write to the Free Software
# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02
#
class FilterColumn < ActiveRecord::Base

  FAMILIES=['date','language','name','links','version','key','metric']

  belongs_to :filter
  validates_inclusion_of :sort_direction, :in => %w( ASC DESC ), :allow_nil => true
  
  def self.create_from_string(string)
    if FAMILIES.include?(string)
      FilterColumn.new(:family => string)
    else
      metric=Metric.by_key(string)
      metric ? FilterColumn.new(:family => 'metric', :kee => metric.key) : nil
    end
  end
  
  def key
    kee
  end

  def name
    if on_metric?
      Api::Utils.message("metric.#{kee}.name", :default => (metric ? metric.short_name : kee))
    else
      Api::Utils.message("filters.col.#{family}", :default => kee)
    end
  end

  def display_name
    name
  end

  def metric
    @metric ||=
      begin
        on_metric? ? Metric.by_key(kee) : nil
      end
  end

  def on_language?
    family=='language'
  end

  def on_version?
    family=='version'
  end

  def on_name?
    family=='name'
  end

  def on_date?
    family=='date'
  end

  def on_links?
    family=='links'
  end

  def on_metric?
    family=='metric'
  end

  def on_profile?
    family=='profile'
  end

  def on_key?
    family=='key'
  end

  def deletable?
    !on_name?
  end

  def sorted?
    sort_direction=='ASC' || sort_direction=='DESC'
  end

  def ascending?
    sort_direction=='ASC'
  end

  def ascending2?
    sort_direction=='ASC'
  end

  def descending?
    sort_direction=='DESC'
  end

  def ascending=(asc)
    write_attribute(:sort_direction, (asc ? 'ASC' : 'DESC'))
  end

  def sortable?
    !on_links?
  end

  def small_width?
    on_metric? && metric && (metric.val_type==Metric::VALUE_TYPE_LEVEL || metric.val_type==Metric::VALUE_TYPE_BOOLEAN) 
  end
end

File: sonar-plugin-api/src/main/java/org/sonar/api/web/FilterColumn.java
/*
 * SonarQube, open source software quality management tool.
 * Copyright (C) 2008-2014 SonarSource
 * mailto:contact AT sonarsource DOT com
 *
 * SonarQube is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * SonarQube is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
package org.sonar.api.web;

import com.google.common.collect.ImmutableSortedSet;

import java.util.Set;

import com.google.common.base.Preconditions;

/**
 * Definition of a {@link Filter} column.
 *
 * @since 3.1
 */
public final class FilterColumn {
  public static final String ASC = "ASC";
  public static final String DESC = "DESC";
  public static final Set<String> DIRECTIONS = ImmutableSortedSet.of(ASC, DESC);

  private final String family;
  private final String key;
  private final String sortDirection;
  private final boolean variation;

  private FilterColumn(String family, String key, String sortDirection, boolean variation) {
    Preconditions.checkArgument(DIRECTIONS.contains(sortDirection), "Valid directions are %s, not '%s'", DIRECTIONS, sortDirection);

    this.family = family;
    this.key = key;
    this.sortDirection = sortDirection;
    this.variation = variation;
  }

  /**
   * Creates a new {@link FilterColumn}.
   *
   * <p>Valid values for the {@code sortDirection} are {@code #ASC}, {@code #DESC}</p>
   *
   * <p>When the @{see Filter} is persisted, a validation is made on the {@code family} and the {@code key}.
   * They should point to a valid column description.</p>
   *
   * @throws IllegalArgumentException if {@code sortDirection} is not valid
   */
  public static FilterColumn create(String family, String key, String sortDirection, boolean variation) {
    return new FilterColumn(family, key, sortDirection, variation);
  }

  /**
   * Get the the column's family.
   * 
   * @return the family
   */
  public String getFamily() {
    return family;
  }

  /**
   * Get the the column's key.
   * 
   * @return the key
   */
  public String getKey() {
    return key;
  }

  /**
   * Get the the column's sort direction.
   * 
   * @return the sort direction
   */
  public String getSortDirection() {
    return sortDirection;
  }

  /**
   * A column can be based on the variation of a value rather than on the value itself.
   * 
   * @return <code>true</code> when the variation is used rather than the value
   */
  public boolean isVariation() {
    return variation;
  }
}
